internal/dataplane/parser/translate_ingress.go

```diff
@@ -39,68 +40,125 @@ func (p *Parser) ingressRulesFromIngressV1beta1() ingressRules {
 				}
-				if host != "" {
-					hosts := kong.StringSlice(host)
-					r.Hosts = hosts
+
+				if routableServices[serviceName][host] == nil {
+					routableServices[serviceName][host] = make(map[int32]map[string]networkingv1beta1.PathType)
 				}
 
-				serviceName := ingress.Namespace + "." +
-					rule.Backend.ServiceName + "." +
-					rule.Backend.ServicePort.String()
-				service, ok := result.ServiceNameToServices[serviceName]
-				if !ok {
-					service = kongstate.Service{
-						Service: kong.Service{
-							Name: kong.String(serviceName),
-							Host: kong.String(rule.Backend.ServiceName +
-								"." + ingress.Namespace + "." +
-								rule.Backend.ServicePort.String() + ".svc"),
-							Port:           kong.Int(DefaultHTTPPort),
-							Protocol:       kong.String("http"),
-							Path:           kong.String("/"),
-							ConnectTimeout: kong.Int(DefaultServiceTimeout),
-							ReadTimeout:    kong.Int(DefaultServiceTimeout),
-							WriteTimeout:   kong.Int(DefaultServiceTimeout),
-							Retries:        kong.Int(DefaultRetries),
+				port := pathRule.Backend.ServicePort.IntVal
+				if routableServices[serviceName][host][port] == nil {
+					routableServices[serviceName][host][port] = make(map[string]networkingv1beta1.PathType)
+				}
+
+				pathType := networkingv1beta1.PathTypePrefix
+				if pathRule.PathType != nil {
+					pathType = *pathRule.PathType
+				}
+				routableServices[serviceName][host][port][path] = pathType
+			}
+		}
+
+		for serviceName, hostnameRules := range routableServices {
+			for host, portRules := range hostnameRules {
+				for portNumber, pathsMap := range portRules {
+					portDef := PortDefFromIntStr(intstr.FromInt(int(portNumber)))
+					kongServiceName := ingress.Namespace + "." + serviceName + "." + portDef.CanonicalString()
+					kongStateService, ok := result.ServiceNameToServices[serviceName]
+					if !ok {
+						kongStateService = kongstate.Service{
+							Service: kong.Service{
+								Name:           kong.String(kongServiceName),
+								Host:           kong.String(fmt.Sprintf("%s.%s.%d.svc", serviceName, ingress.Namespace, portNumber)),
+								Port:           kong.Int(DefaultHTTPPort),
+								Protocol:       kong.String("http"),
+								Path:           kong.String("/"),
+								ConnectTimeout: kong.Int(DefaultServiceTimeout),
+								ReadTimeout:    kong.Int(DefaultServiceTimeout),
+								WriteTimeout:   kong.Int(DefaultServiceTimeout),
+								Retries:        kong.Int(DefaultRetries),
+							},
+							Namespace: ingress.Namespace,
+							Backends: []kongstate.ServiceBackend{{
+								Name:    serviceName,
+								PortDef: portDef,
+							}},
+						}
+					}
+
+					var routeName string
+					if host == "" {
+						routeName = fmt.Sprintf("%s.%s.%d", ingress.Namespace, ingress.Name, portNumber)
+					} else {
+						routeName = fmt.Sprintf("%s.%s.%s.%d", ingress.Namespace, ingress.Name, host, portNumber)
+					}
```

Comment:
Do we not need to incorporate the service name into these? Although these get nested under the service object and will thus route correctly, the names are still independent. Does this not collide on something like

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: example
  namespace: spearmint
spec:
  ingressClassName: nginx-example
  rules:
  - host: "example.com"
    http:
      paths:
      - pathType: Prefix
        path: "/foo"
        backend:
          service:
            name: service1
            port:
              number: 80
      - pathType: Prefix
        path: "/bar"
        backend:
          service:
            name: service2
            port:
              number: 80
      - pathType: Prefix
        path: "/baz"
        backend:
          service:
            name: service3
            port:
              number: 8000
```

Since we have three services, we need to generate two routes, and will get:

```
spearmint.example.example.com.80
spearmint.example.example.com.80
spearmint.example.example.com.8000
```

Reply:
Whoops you're right I overlooked this.

Reply:
b2be34d7e6bb40e273a8f7f6abc3a1cb7a3acab9